Added support for artist releases endpoint

maggiolo00 opened the pull request:
Hi @afonso360 
i've created a first support for artist releases.
I did create this [repo](https://github.com/RustRome/discography) a couple of days but then i saw this, and it's better to contribute to your repo and join the forces :D

Tell me what you think, 
Thanks
Enrico

afonso360 commented on src/data_structures/others.rs:
```diff
@@ -81,3 +81,19 @@ pub enum DataQuality {
+
+
+#[derive(Deserialize, Debug)]
+pub struct Pagination {
```
Move this to /src/pagination.rs please

afonso360 commented on src/data_structures/artist.rs:
````diff
@@ -152,6 +152,20 @@ impl ArtistQueryBuilder {
+    /// let releases = Discogs::new(env!("DISCOGS_USER_AGENT"))
+    ///                       .artist(1234).releases();
+    /// ```
+    pub fn releases(&self) -> ArtistReleasesQueryBuilder {
````
If we do end up doing a `get_releases` function we could maybe set an internal flag here so that `get_query_url` knows to build a releases url instead of the regular one

maggiolo00 commented:
hi @afonso360 

I like this approach with nested builder but we   could try to not use it and use a
`get_releases` function, but it  should take pagination parameters and also sort paramenter

https://www.discogs.com/developers/#page:database,header:database-artist-releases

About pagination yes we should find a solution that is reusable across endpoints

i've tried to use `Releases` struct but seems that is different 

https://api.discogs.com/artists/4567/releases?page=1&per_page=2
vs
https://api.discogs.com/releases/128

afonso360 commented:
> I like this approach with nested builder but we   could try to not use it and use a
> `get_releases` function, but it  should take pagination parameters and also sort paramenter
Yeah! Altough, in terms of api would this look like
```
Discogs::new()
	.artist(123)
	.get_releases(Pagination {
		per_page: 10,
		page: 1
	}, Sort {
		type: SortType::Year,
		order: SortOrder::Ascending
	})
```

or

```
Discogs::new()
	.artist(123)
	.pagination(Pagination {
		per_page: 10,
		page: 1
	})
	.sort(Sort {
		type: SortType::Year,
		order: SortOrder::Ascending
	})
	.get_releases()
```

I think the second one represents the discogs api better just because
those fields are optional in the api, however it has the problem
where a user might call them and then call `get()` which would
have no effect at all. And might be confusing.


> i've tried to use `Releases` struct but seems that is different 
> 
> https://api.discogs.com/artists/4567/releases?page=1&per_page=2
> vs
> https://api.discogs.com/releases/128

The reason i think they are the same is that the releases from the first
one point to the second one in their resource_url, and have mostly the
same fields

```
id
resource_url
status
thumb
title
year
```

The ones that are diffrent are just a diffrent representation of the
same data.

For example, the `artists` and `role` field on the artist release data are
inside the `artists` field on the full release data, this makes
sense because the release can have multiple artists, however when you
search for that artists releases it shows any releases that he
contributed in.

I think it would make more sense to just have one release structure and
use that across the whole library even if it diverges from the API a
little bit, what do you think about this?



The reason that I would prefer to have one release structure is mostly
for the way that im planing on addressing the issue of incomplete data
such as when you for example request

https://api.discogs.com/releases/843401

And iterate trough the artists field, that is clearly an `Artist` data
structure, however its incomplete, so we could have an `update()` method
to fetch the complete data and fill in the blanks, if we have different
data structures this sort of stops making sense.

Anyway, this is something to be solved in the future and not now.

afonso360 commented:
> I like this approach with nested builder but we   could try to not use it and use a
> `get_releases` function, but it  should take pagination parameters and also sort paramenter
Yeah! Altough, in terms of api would this look like
```
Discogs::new()
	.artist(123)
	.get_releases(Pagination {
		per_page: 10,
		page: 1
	}, Sort {
		type: SortType::Year,
		order: SortOrder::Ascending
	})
```

or

```
Discogs::new()
	.artist(123)
	.pagination(Pagination {
		per_page: 10,
		page: 1
	})
	.sort(Sort {
		type: SortType::Year,
		order: SortOrder::Ascending
	})
	.get_releases()
```

I think the second one represents the discogs api better just because
those fields are optional in the api, however it has the problem
where a user might call them and then call `get()` which would
have no effect at all. And might be confusing.


> i've tried to use `Releases` struct but seems that is different 
> 
> https://api.discogs.com/artists/4567/releases?page=1&per_page=2
> vs
> https://api.discogs.com/releases/128

The reason i think they are the same is that the releases from the first
one point to the second one in their resource_url, and have mostly the
same fields

```
id
resource_url
status
thumb
title
year
```

The ones that are diffrent are just a diffrent representation of the
same data.

For example, the `artists` and `role` field on the artist release data are
inside the `artists` field on the full release data, this makes
sense because the release can have multiple artists, however when you
search for that artists releases it shows any releases that he
contributed in.

I think it would make more sense to just have one release structure and
use that across the whole library even if it diverges from the API a
little bit, what do you think about this?



The reason that I would prefer to have one release structure is mostly
for the way that im planing on addressing the issue of incomplete data
such as when you for example request

https://api.discogs.com/releases/843401

And iterate trough the artists field, that is clearly an `Artist` data
structure, however its incomplete, so we could have an `update()` method
to fetch the complete data and fill in the blanks, if we have different
data structures this sort of stops making sense.

Anyway, this is something to be solved in the future and not now.

maggiolo00 commented:
hi @afonso360 

i like this one too
```
Discogs::new()
	.artist(123)
	.pagination(Pagination {
		per_page: 10,
		page: 1
	})
	.sort(Sort {
		type: SortType::Year,
		order: SortOrder::Ascending
	})
	.get_releases()
```

but i get that can confuse since this can happen

```
Discogs::new()
	.artist(123)
	.pagination(Pagination {
		per_page: 10,
		page: 1
	})
	.sort(Sort {
		type: SortType::Year,
		order: SortOrder::Ascending
	})
	.get();
```


I Agree we should try to use the same struct Release maybe change it a little bit to fit also

Artist/Label releases

afonso360 commented:
Yeah, it might have to be something that we document and hope people understand.

Anyway, I think we have enough to start working on this, lets do it!

maggiolo00 commented:
Sounds good :D

maggiolo00 commented:
hi @afonso360 

i've implemented the `get_releases` function and changed a little bit the `Release` struct in order to use it also in ArtistReleases endpoint.

With this impl i think only one drawback that is the possibility to do this

```rust
 let mut artist = Discogs::new(env!("DISCOGS_USER_AGENT"))
                    .artist(4567);

                artist
                .pagination(1, 2)
                .get_releases();

                artist.get();
```

but the second call will fails since it's configured for fetching releases.
I guess we could use consuming builders to prevent that

https://aturon.github.io/ownership/builders.html#consuming-builders:

afonso360 commented:
Looks good, I think we should address that issue in a separate issue since that pattern repeats across the whole api.

Thank you for taking the time to build this pull request
